noch ein paar Änderungen + Anmerkungen

maxbachmann opened the pull request:
Die entfernte Funktion kann unter keinen Umständen aufgerufen werden, da m_x immer nur erhöht oder auf 0 gesetzt wird. dh Sie ist nie unter null
Das selbe gilt in Zeile 101 für die or Bedingung die ebenfalls hinfällig ist

gegenstrecke wird lediglich verringert aber nie erhöht also ist eine Prüfung nach oben aus Zeile 109 hinfällig

Nachdem die Variablen beide nie unter 0 sind und danach immer als uint8_t übergeben werden kann man sie auch von Anfang an so definieren. Auch für gegenstrecke bietet es sich an auf uint8_t umzusteigen.

m_x kann wohl maximal 6 werden da bei m_x > 5 = 0 gesetzt wird
xA = 5-m_x -> kann -1 sein wird danach dann in Zeile 115 als uint8_t übergeben ist das beabsichtigt? das wäre dann der maximale Wert den eine uint8_t annehmen kann was mir irgendwie falsch erscheint


in zeile 114 wird hue erhöht -> in Zeile 115 und 116 ist hue potentiell 256 da die Prüfung erst später erfolgt. Ist das beabsichtigt? das klingt für mich mal falsch

Ich bin mir gerade nicht mehr ganz sicher aber ich glaube die Erhöhung von hue in Zeile 115 erfolgt erst nach dem Funktionsaufruf ist das so erwünscht?

maxbachmann committed:
changed variable name

lokale Variablen sollten nicht den gleichen Namen wie globale haben (auch wenn das in diesem Fall trotzdem funktioniert)

maxbachmann committed:
changed variable type

removed some complexity

maxbachmann commented:
btw zu der Anmerkung bei Zeile 114 nachdem uint8_t ja ein byte ist und damit maximal 255 springt der Wert danach mit dem nächsten ++ wieder auf 0 dementsprechend brauchst du nicht danach auf >255 testen das kann nie passieren und auf 0 wird sie eh von allein gesetzt ;)

maxbachmann committed:
removed unused if

this if could never be true
and changed variable name from i to returnValue, since i is a variablename you might use for a counter but it does not really give a good description of what it´s used for

gegenstrecke is always between 0 and a value smaller than 255 and is passed as uint8_t so it can be changed to unit8_t same with xB and xA they get passed as uint8_t so they have to be in this range anyways

maxbachmann commented:
Also als mögliche Probleme die noch bestehen wo ich aber nicht weiß was der Gedanke war:

die Erhöhung von hue in dem Funktionsaufruf als post increment was schätzungsweise dafür sorgt dass nicht alle drei Funktionsaufrufe den selben Wert für hue bekommen

die Erhöhung von m_x in Zeile 93 die dafür sorgen kann dass xA 255 wird, was schätzungsweise unerwünscht ist

DEvil0000 commented on toolboxTreppe/toolboxTreppe.ino:
```diff
@@ -93,23 +93,22 @@ void animation1() {
       onewayMillis = currentMillis;
-      --gegenstrecke;
-      if (gegenstrecke < 0 || gegenstrecke > kMatrixHeight-1) {
+      if (gegenstrecke == 0) {
```
das scheint mir falsch -> !=

maxbachmann replied:
wieso der aktuelle Check ist ist Gegenstrecke < 0 oder > kMatrixheight-1
es kann aber nie größer als kmatrixHeight-1 sein da nur --gegenstrecke gemacht wird
also ist der check nach oben hin schonmal sinnlos

der check nach unten wurde ersetzt durch ==0 da immer -1 gemacht wird, das jetzt aber erst nach diesem check gemacht wird.
Dh in dem Durchlauf in dem es bisher auf -1 gestellt wurde und auf kmatrixHeight -1 wird es jetzt einfach direkt auf kmatrixheight-1 gesetzt oder sehe ich das falsch?

DEvil0000 replied:
ja richtig

DEvil0000 commented on toolboxTreppe/toolboxTreppe.ino:
```diff
@@ -93,23 +93,22 @@ void animation1() {
+      if (gegenstrecke == 0) {
         gegenstrecke = kMatrixHeight-1;
+      } else {
+        --gegenstrecke;
```
hier würd ich auch nochmal hingucken

maxbachmann replied:
das ist in der Nachricht darüber erklärt 
nachdem gegenstrecke jetzt auch eine uint8_t ist muss das auch sorum erfolgen, da sie ja nicht mehr negativ werden kann (könnte natürlich auch auf 255 gecheckt werden) 
aber man spart sich so in dem Durchlauf in dem auf kmatrixheight -1 gesetzt wird das --gegenstrecke

DEvil0000 commented on toolboxTreppe/toolboxTreppe.ino:
```diff
@@ -116,9 +116,6 @@ void animation1() {
     leds[ XY(5, gegenstrecke)] = CHSV( hue, 255, 255 );
     FastLED.show();
-    if (hue > 255) {
-      hue = 0;
```
hue wird in 115 erhöht. reset nötig

maxbachmann replied:
hue ist eine uint8_t das heißt nach 255 kommt 0 
weiterhin kann hue niemals größer als 255 sein und dementsprechend der check nie etwas machen

maxbachmann replied:
btw bezüglich 115 weißt du zufällig ob das post increment da dann schon den erhöhten Wert weitergibt? ich dachte das ist dann erst für die nächste Funktion der erhöhte Wert (wobei ich natürlich auch nicht sicher weiß was da an der Stelle das Ziel ist daher habe ich das auch mal so gelassen)

DEvil0000 replied:
hast recht uint8 überlauf ist mir entgangen.
post inc macht wie der name schon sagt das inc hinterher bzw returnt den wert vor dem inc. vermutlich soll es hier ein pre sein

maxbachmann replied:
ok dachte ich mir auch ;)

maxbachmann replied:
vermutlich wäre es einfach sinnvoller das Inkrement dann aber davor oder danach zu stellen.
Das in der Übergabe zu machen ist eher unübersichtlich (und mit pre/post inkrement tendenziell auch fehleranfälliger)

DEvil0000 commented:
für die übersicht hätte ich es auch davor gestellt.

maxbachmann committed:
change order

changed the order to prevent xA from overflowing

hue gets now incremented outside the CHSV function call

maxbachmann commented:
Ok ich habe das dann noch dementsprechend angepasst und auch das erwähnte Problem mit xA behoben bei dem es zu einem Überlauf kommt

DEvil0000 commented:
localizing variables would be a good improvement.

maxbachmann commented:
stimmt mache ich später ;)

maxbachmann commented:
Was ist der generelle Ton bei Arduino werden da so global Sachen wie ein Pin define
wie hier mit const int oder #define gelöst?
Ich hätte da jetzt persönlich eher #define genutzt

maxbachmann committed:
made all variables local

replaced global definitions by defines

maxbachmann commented:
Das sollte funktionieren
bei der debug() bin ich mir nicht sicher ob vom performance Standpunkt bei etwas so kleinem wie uint8_t das kopieren vielleicht schneller wäre als das dereferenzieren (meine mal etwas entsprechendes mitbekommen zu haben)

maxbachmann committed:
removed #define that was double

maxbachmann committed:
added definitions for functions in top

now using const reference for xY and XY safe aswell
returnvalue will not be changed -> const

maxbachmann commented:
btw xysafe wird als Funktion nicht genutzt

maxbachmann commented:
@kr0l benötigst du den debug code immer?
mein Ansatz ist da in der Regel eher ein
#define DEBUG
und dann #ifdef DEBUG um die entsprechenden Funktionen, so dass der debug code nur ausgegeben wird in der Entwicklungsphase und danach dann durch auskommentieren von #define DEBUG in der End Version nicht vorhanden ist (natürlich im Falle von Problemen schnell wieder aktiviert ;) )

DEvil0000 commented:
define oder const:
am ende egal

kopieren vs deref:
hängt vom cpu und deren instuction set und der compiler optimierung ab. hat beides vor und nachteile. kopieren kann z.b. im cpu register liegen und damit schneller sein.
bei dem code aber egal

funktionen die nicht verwendet werden:
löschen

define debug:
hat auch vor und nachteile. du änderst damit das laufzeit verhalten und timing. außerdem kannst du dann debug nicht während der laufzeit z.b. per switch an und aus schalten.

maxbachmann commented:
Ok ja wärend der Laufzeit ist dann in schalten nicht möglich (benötige ich in der Regel nicht daher die Nachfrage)

Wie löst du dann debugausgaben?
Meine Lösung ist in der Regel

#define DEBUG

#ifdef DEBUG
  #define DEBUG_PRINT(x)    Serial.print(x)
  #define DEBUG_PRINTLN    Serial.println(x)
#else
  #define DEBUG_PRINT(x)
  #define DEBUG_PRINTLN(x)
#endif

DEvil0000 commented on toolboxTreppe/toolboxTreppe.ino:
```diff
@@ -56,25 +46,20 @@ void setup() {
     //Wenn der Schalter in der mitte ist, aktiviere debugfunktion
     if (digitalRead(buttonPin2) == HIGH) {
-      debug();
+      debug(y, x, gegenstrecke);
```
just enable debug here

DEvil0000 commented on toolboxTreppe/toolboxTreppe.ino:
```diff
@@ -86,43 +71,49 @@ void loop()
 //-------------------------------Animationsfunktion ------------
-void animation1() {
-  currentMillis = millis();
+void animation1(uint8_t& y, uint8_t& x, uint8_t& gegenstrecke) {
```
then x, y and gegenstrecke are local to this function

DEvil0000 commented on toolboxTreppe/toolboxTreppe.ino:
```diff
@@ -86,43 +71,49 @@ void loop()
+  leds[ XY(xA, y)] = CHSV( hue, 255, 255 );
+  leds[ XY(xB, y)] = CHSV( hue, 255, 255 );
+  leds[ XY(5, gegenstrecke)] = CHSV( hue, 255, 255 );
+  FastLED.show();
```
call debug e.g. here

DEvil0000 commented on toolboxTreppe/toolboxTreppe.ino:
```diff
@@ -86,43 +71,49 @@ void loop()
 //----------------------------- Debug Funktion ----------------
-void debug() {
-  Serial.print(m_y);
+void debug(const uint8_t& y, const uint8_t& x, const uint8_t& gegenstrecke) {
```
debug could be a generic print function like sprintf

maxbachmann replied:
the debug function already makes no sence in the current way anyways.
He activates debug when pin1 and 2 are both high and then in the debugfunction 
so the debug output Oben and Unten can´t appear and the output Mitte will always appear

DEvil0000 replied:
it's cleanup time

kr0l replied:
I wrote the debug function before i planned how to acces it and then i forgot to remove the section with the button modes...

DEvil0000 commented on toolboxTreppe/toolboxTreppe.ino:
```diff
@@ -146,8 +141,13 @@ void powerOff() {
+}
 
-uint16_t XY(const uint8_t x, const uint8_t y) {
+const uint16_t XY(const uint8_t& x, const uint8_t& y) {
```
XY could be a struct with constructor

maxbachmann replied:
hm I never used structs with a constructor yet, but what would be the advantage when using them instead of the XY function?

DEvil0000 replied:
basically none - just style

maxbachmann replied:
Hm I am still not quite sure how you mean this
I mean basically XY calculates a pin number so when I understand you right you would prefer a struct
that includes the variable pinNumber and the constructor of the struct gets called with the parameters x and y, calculates the pin number and sets the pinNumber variable to this value
Then you would have to access this value.

In comparision the current way you call the function and it returns the pin that you need at this place
I mean you only need this pin once, so I can´t really see a big advantage in using a struct even in terms of style, or is my idea on how you would implement this struct

DEvil0000 replied:
kanns dir einfacher am do mal erklären

kr0l commented:
Cool, da ist ja echt viel passiert. Danke. 

Der Debug Code muss nicht immer auf schalterdruck aufrufbar sein, da im Normalfall die Treppe funktioniert und wenn man debuggen muss ein Neustart kein Problem sein sollte.

maxbachmann committed:
changed debugfunction

maxbachmann commented:
I changed the debugfunction a bit. I guess thats what @DEvil0000 proposed, obviously more functions could be added to the debugfunction, but I guess this one is enough since there is only one debug output now anyways

for the struct part I am really not sure what you mean exactly: see my answer to your improvement request

DEvil0000 commented:
ich glaub deine debug funktion wird so nicht richtig funktionieren.

DEvil0000 commented on toolboxTreppe/toolboxTreppe.ino:
```diff
@@ -110,43 +105,20 @@ void animation1(uint8_t& y, uint8_t& x, uint8_t& gegenstrecke) {
-  if (digitalRead(buttonPin4) == LOW) {
-    Serial.print("Knopf 4");
+
+void debugPrint(const char* output[]) {
```
your debug funktion will not do the thing you want.

DEvil0000 replied:
have a look here: http://www.cplusplus.com/reference/cstdio/sprintf/

DEvil0000 commented on toolboxTreppe/toolboxTreppe.ino:
```diff
@@ -43,40 +42,37 @@ void setup() {
+  debugPrint(
+    "x: " + x + "   "
+    + "y: " + y + "   "
+    + "gegenstrecke: " + gegenstrecke
```
like this you read them as chars which will not work out well and is not readable.

maxbachmann commented:
ok lass uns da am do mal durchsprechen 
dann bin ich danach auch mal wieder ein wenig klüger :+1:

maxbachmann committed:
fixed the debug Function

maxbachmann commented:
Ich denke mal so sollte die Debug Funktion jetzt funktionieren. Kannst ja vor dem mergen den code mal noch kurz auf dem Arduino testen

DEvil0000 commented on toolboxTreppe/toolboxTreppe.ino:
```diff
@@ -30,7 +31,7 @@ CRGB leds_plus_safety_pixel[ NUM_LEDS + 1];
 
 void animation1(uint8_t& y, uint8_t& x, uint8_t& gegenstrecke);
-void debugPrint(const char* output);
+void debugPrint(const char* output[], ...);
```
I think this has to be void debugPrint(const char* output, ...);

DEvil0000 commented on toolboxTreppe/toolboxTreppe.ino:
```diff
@@ -106,9 +103,10 @@ void animation1(uint8_t& y, uint8_t& x, uint8_t& gegenstrecke) {
 //----------------------------- Debug Funktion ----------------
 
-void debugPrint(const char* output[]) {
+void debugPrint(const char* output[], ...) {
```
void debugPrint(const char* output, ...);

maxbachmann replied:
already changed on thursday after we talked but not pushed yet

DEvil0000 commented on toolboxTreppe/toolboxTreppe.ino:
```diff
@@ -106,9 +103,10 @@ void animation1(uint8_t& y, uint8_t& x, uint8_t& gegenstrecke) {
+  va_list args;
   if (digitalRead(buttonPin1) && digitalRead(buttonPin2)) {
-    Serial.println(output);
+    Serial.println(output, args);
```
and here its wrong.
Serial.println takes 2 params: 1) the string 2) a "mode" like print in hex
you need to call sprintf() to get the final string and pass it to serial.println

maxbachmann replied:
I know already saw that in the error messages on thursday but had no time to change it yet

DEvil0000 commented on toolboxTreppe/toolboxTreppe.ino:
```diff
@@ -130,5 +128,4 @@ const uint16_t XY(const uint8_t& x, const uint8_t& y) {
   return returnValue;
-}
-
+}
```
there should be an empty line at the end

maxbachmann replied:
already changed on thursday but not pushed yet

DEvil0000 commented on toolboxTreppe/toolboxTreppe.ino:
```diff
@@ -43,40 +42,37 @@ void setup() {
-
-
 //-------------------------------Animationsfunktion ------------
 void animation1(uint8_t& y, uint8_t& x, uint8_t& gegenstrecke) {
```
function does not take any arugments anymore - they are local to it

maxbachmann replied:
already changed on thursday but not pushed yet

maxbachmann commented:
Ich weiß die Fehler sind mir schon bekannt (sind lokal auch schon größtenteils abgeändert) 
Ich habe auch hinter den ganzen defines ein ; das da nicht hingehört ^^

maxbachmann committed:
fixed a couple syntax issues and added sprintf

maxbachmann committed:
fixed end of file

maxbachmann commented:
ka wieso aber in visual studio code wird irgendwie der letzte new line nicht mitgezählt :/ 
da brauchts immer 2 new line damit einer in der datei ankommt  (microsoft denkt halt mit xD)

maxbachmann commented:
jedenfalls beim kompilierversuch hier bei mir kommen jetzt nur noch Fehler die die LED library betreffen, da ich die nicht vorliegen habe

maxbachmann committed:
forgot to change a name

maxbachmann committed:
forgot to change the parameters in the declaration

DEvil0000 commented on toolboxTreppe/toolboxTreppe.ino:
```diff
@@ -103,10 +103,11 @@ void animation1(uint8_t& y, uint8_t& x, uint8_t& gegenstrecke) {
   va_list args;
-  if (digitalRead(buttonPin1) && digitalRead(buttonPin2)) {
-    Serial.println(output, args);
+  char* buffer;
```
has to be a allocated buffer:
char buffer[1024]; //size of 1024 is not too small I hope

maxbachmann replied:
true will change that and yes I guess nobody should print more than 1024 chars per line when he wants a readable debug output

DEvil0000 commented on toolboxTreppe/toolboxTreppe.ino:
```diff
@@ -30,7 +30,7 @@
 CRGB* const leds( leds_plus_safety_pixel + 1);
 
-void animation1(uint8_t& y, uint8_t& x, uint8_t& gegenstrecke);
+void animation(uint8_t& y, uint8_t& x, uint8_t& gegenstrecke);
```
it does not have parameters anymore

maxbachmann committed:
fixed char array

DEvil0000 commented:
sieht für mich jetzt gut aus. @kr0l frag bei gelegenheit doch mal den compiler und pack die version auf die treppe
edit: merge bitte als squash - sonst ist die history so vollgemüllt
